cmd/kubeadm/app/images/interface.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package images
+
+import (
+	"fmt"
+
+	kubeadmapiv1alpha2 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2"
+	utilsexec "k8s.io/utils/exec"
+)
+
+// Puller is an interface for pulling images
+type Puller interface {
+	Pull(string) error
+}
+
+// Existence is an interface to determine if an image exists on the system
+// A nil error means the image was found
+type Existence interface {
+	Exists(string) error
+}
+
+// Images defines the set of behaviors needed for images realting to the CRI
+type Images interface {
+	Puller
+	Existence
+}
+
+// CRInterfacer is a struct that interfaces with the container runtime
+type CRInterfacer struct {
+	criSocket  string
+	exec       utilsexec.Interface
+	crictlPath string
+	dockerPath string
+}
+
+// NewCRInterfacer sets up and returns a CRInterfacer
+func NewCRInterfacer(execer utilsexec.Interface, criSocket string) (*CRInterfacer, error) {
+	var crictlPath, dockerPath string
+	var err error
+	if criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
+		if crictlPath, err = execer.LookPath("crictl"); err != nil {
+			return nil, fmt.Errorf("crictl is required for non docker container runtimes: %v", err)
+		}
+	}
+	if criSocket == kubeadmapiv1alpha2.DefaultCRISocket {
+		if dockerPath, err = execer.LookPath("docker"); err != nil {
+			return nil, fmt.Errorf("`docker` is required when docker is the container runtime: %v", err)
+		}
+	}
+
+	return &CRInterfacer{
+		exec:       execer,
+		criSocket:  criSocket,
+		crictlPath: crictlPath,
+		dockerPath: dockerPath,
+	}, nil
+}
+
+// Pull pulls the actual image using either crictl or docker
+func (cri *CRInterfacer) Pull(image string) error {
+	if cri.criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
+		return cri.exec.Command(cri.crictlPath, "-r", cri.criSocket, "pull", image).Run()
+	}
+	return cri.exec.Command(cri.dockerPath, "pull", image).Run()
+}
+
+// Exists checks to see if the image exists on the system already
+// Returns an error if the image is not found.
+func (cri *CRInterfacer) Exists(image string) error {
+	if cri.criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
```

Comment:
@luxas @stealthybox I want to draw your attention to this code as I'm now making the assumption that any failure of `crictl inspecti` or `docker inspect` means the image does not exist.

We know the binary exists because if we need the binary, we make sure it exists before returning the `CRInterfacer`.

The  other option is doing Stderr/Stdout parsing and that is a guess (albeit we could be reasonably smart about it).

Reply:
@chuckha I think passing up whatever the error is when the commands fail sounds fine.
This if statement could probably use a little hardening though:
```golang
if cri.criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
  // cri -r <socket> inspecti
  // this if statement assumes that cri.crictlPath exists -- should probably be more explicit
} else if cri.crictlPath != "" {
  // this case looks like it's missing? currently falls back to docker if default CRI socket
  // a comment should probably be here to explain that this is expected
} else if cri.dockerPath != "" {
  // docker inspect
}
// should return friendly error message here that we don't have a supported image checker and that the user can skip this check
```

Reply:
We do a check for the binary existence in the NewCRInterfacer function.

Filling out the snippet would look like this:

```go
if cri.criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
  // cri -r <socket> inspecti
} else if cri.crictlPath != "" {
  // cri -r <socket> inspecti
} else if cri.dockerPath != "" {
  // docker inspect
}
// should return friendly error message here that we don't have a supported image checker and that the user can skip this check
```

One thing to note is that the `cri.exec.Command` will return a specific `exec.ErrExecutableNotFound` if the binary is not found that will get bubbled up to the user. I could see an argument for wrapping it, but also we do that check already in the constructor so I'm not sure it makes sense to do it again.

Reply:
After the small change to the interfacer constructor, I understand the code-path much better.
I'm cool with this :)